torch_frame/nn/encoder/stype_encoder.py

```diff
@@ -685,3 +730,86 @@ def encode_forward(self, feat: TensorData) -> Tensor:
         # -> [batch_size, num_cols, out_channels]
         x = x_lin + self.bias
         return x
+
+
+class TimestampEncoder(StypeEncoder):
+    r"""TimestampEncoder for timestamp stype. The encoder applies
+    different frequencies to the timestamps to calculate the
+    positional encoding and then adds the positional encoding to
+    the original :obj:`Tensor`. For year, we use the positional
+    encoding introduced in `"Attention Is All You Need"
+    <https://arxiv.org/abs/1706.03762>`_ paper. For month, day,
+    dayofweek, hour, minute and second, we use alternating
+    sine and cosine of (2 * pi * x, 4 * pi * x, ...) up to num_freqs.
+    We add the positional encoding to the original tensor and then
+    use a :class:`~torch.nn.Linear` layer to map back to out_channels.
+
+    Args:
+        num_freqs: Number of times to apply the different sine and
+        cosine frequencies.
+    """
+    supported_stypes = {stype.timestamp}
+
+    def __init__(
+        self,
+        out_channels: Optional[int] = None,
+        stats_list: Optional[List[Dict[StatType, Any]]] = None,
+        stype: Optional[stype] = None,
+        post_module: Optional[Module] = None,
+        na_strategy: Optional[NAStrategy] = None,
+        num_freqs: int = 8,
+    ):
+        self.num_freqs = num_freqs
+        super().__init__(out_channels, stats_list, stype, post_module,
+                         na_strategy)
+
+    def init_modules(self):
+        super().init_modules()
+        min_year = torch.tensor([
+            self.stats_list[i][StatType.YEAR_RANGE][0]
+            for i in range(len(self.stats_list))
+        ])
+        self.register_buffer("min_year", min_year)
+        div_term = torch.exp(
+            torch.arange(0, self.num_freqs, 2).float() *
+            (-np.log(10000.0) / self.num_freqs))
+        self.register_buffer("div_term", div_term)
+        mult_term = 2 * torch.pi * torch.arange(
+            1, (self.num_freqs + 1) // 2 + 1).float()
+        self.register_buffer("mult_term", mult_term)
+        position = torch.arange(0, stype.timestamp.timestamp_num_cols - 1,
```

Comment:
I think you can replace `stype.timestamp.timestamp_num_cols` here just with `len(TIME_TO_INDEX)`. I think it is quite confusing to expose `timestamp_num_cols` as a property of `stype`.

Reply:
This property is no longer necessary after we make the encoding modular